core/src/test/java/io/grpc/DeadlineTest.java

```diff
@@ -144,23 +149,43 @@ public void run() {
   }
 
   @Test
-  public void testToString() {
+  public void toString_exact() {
     Deadline d = Deadline.after(0, TimeUnit.MILLISECONDS);
-    assertEquals(0, extractRemainingTime(d.toString()), delta);
 
-    d = Deadline.after(-1, TimeUnit.HOURS);
-    assertEquals(d.timeRemaining(TimeUnit.NANOSECONDS), extractRemainingTime(d.toString()), delta);
+    assertDeadlineEquals(d, extractRemainingTime(d.toString()), maxDelta, NANOSECONDS);
+  }
+
+  @Test
+  public void toString_after() {
+    Deadline d = Deadline.after(-1, TimeUnit.HOURS);
+
+    assertDeadlineEquals(d, extractRemainingTime(d.toString()), maxDelta, NANOSECONDS);
+  }
+
+  @Test
+  public void toString_before() {
+    Deadline d = Deadline.after(10, TimeUnit.SECONDS);
 
-    d = Deadline.after(10, TimeUnit.SECONDS);
-    assertEquals(d.timeRemaining(TimeUnit.NANOSECONDS), extractRemainingTime(d.toString()), delta);
+    assertDeadlineEquals(d, extractRemainingTime(d.toString()), maxDelta, NANOSECONDS);
+  }
+
+  public static void assertDeadlineEquals(
+      Deadline expected, Deadline actual, long delta, TimeUnit timeUnit) {
+    // This is probably overkill, but easier than thinking about overflow.
+    BigInteger actualTimeRemaining = BigInteger.valueOf(actual.timeRemaining(NANOSECONDS));
+    BigInteger expectedTimeRemaining = BigInteger.valueOf(expected.timeRemaining(NANOSECONDS));
```

Comment:
You are checking `timeRemaining()` against itself. Such test is flawed. With a hypothetical bug in `timeRemaining()` that all return values were off by the same value, `assertDeadlineEquals()` wouldn't be able to catch it.

The result of `timeRemaining()` should be checked against canonical values. I would recommend reverting to the original test pattern and use something like `assertEquals(long, long, long)`.